proxy/stats_handler.go

```diff
@@ -0,0 +1,215 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/websocket"
+
+	"github.com/rancherio/websocket-proxy/common"
+)
+
+type StatsHandler struct {
+	backend         backendProxy
+	parsedPublicKey interface{}
+}
+
+func (h *StatsHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	multiHost := false
+
+	if strings.HasSuffix(req.URL.Path, "project/") || strings.HasSuffix(req.URL.Path, "service/") {
+		multiHost = true
+	}
+
+	hostKeys, urls, authed := h.auth(req, multiHost)
+	if !authed {
+		return
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+	ws, err := upgrader.Upgrade(rw, req, nil)
+	if err != nil {
+		http.Error(rw, "Failed to upgrade connection.", 500)
+		return
+	}
+
+	msgKeys := make([]string, len(hostKeys))
+	respChannels := make([]<-chan common.Message, len(hostKeys))
+
+	var mutex sync.Mutex
+	var countMutex sync.Mutex
+
+	doneCounter := len(hostKeys)
+
+	for i, hostKey := range hostKeys {
+		var err error
+		msgKeys[i], respChannels[i], err = h.backend.initializeClient(hostKey)
+		if err != nil {
+			h.closeConnection(ws)
+			return
+		}
+
+		// Send response messages to client
+		go func(pos int) {
```

Comment:
Perhaps the whole thing would feel cleaner if you pulled this out into a separately declared function (if combined with the above suggestion to create a struct housing all the needed data for a particular backend connection).